### ISSUE TITLE
What is the way to wait until all actions of one type ends?

### ISSUE THREAD

**Issue by bondom** [redux-saga/redux-saga#1390]
I have todo list with possibility to delete items of list, my saga listens on corresponding action:
```
export default function* deleteTodoListItemSaga() {
  while (true) {
    const action = yield take(DELETE_TODO_LIST_ITEM_ACTIONS.REQUEST);
    yield fork(deleteTodoListItem, action);
  }
}
```

I get todo list after item was  successfully deleted:

```
export function* deleteTodoListItem(action: {
  type: string;
  data: { id: number };
}) {
  try {
    const result = yield call(api.deleteTodoListItem, action.data.id);
    yield put(success(result));
    yield put(getTodoList()); // this line
  } catch (e) {
    yield put(err(e));
  }
}
```

Explanation of problem: 
I want to have interactive ui, so user can delete several items at once. 
When todo list is requested from API, loader is shown
So, when user subsequently deletes several items, there is can be such situation: 
 1) first item was successfully deleted(response is 200) => todo list is requested => loader is shown
 2) second item was successfully deleted => todo list is requested again => loader is shown again

So, I see one solving of this problem:
To get todo list from api after all running `deleteTodoListItem` end.
But how can I achieve this?

**Comment by Andarist** [comment]
U need to introduce some temporary array, track completions in it and dispatch `getTodoList` conditionally based on the fact if there is still any `deleteTodoListItem` running.

**Comment by bondom** [comment]
Yes, my problem was that I tried to implement this task with saga and didn't know how to do it.
For task described above this can be easily implemented with deleteTodoListItem reducer and without some special saga watchers.

BUT: suppose that we have possibility to add new todo items:
1) User presses 'Add' button => request for adding new todo is being sent
2) While first request is being sent user presses 'Delete' button on some item => request for deleting todo item is being sent.

We have two separate reducers for adding and deleting items

So to be able to wait until adding and deleting request will finish I created special watcher saga:

```
export default function* getTodoListDueUpdateSaga() {
  let processingRequestsNumber = 0;

  while (true) {
    const action = yield take([
      DELETE_ACTIONS.REQUEST,
      DELETE_ACTIONS.GOT,
      DELETE_ACTIONS.ERROR,
      ADD_ACTIONS.REQUEST,
      ADD_ACTIONS.GOT,
      ADD_ACTIONS.ERROR
    ]);

    switch (action.type) {
      case DELETE_ACTIONS.REQUEST:
      case ADD_ACTIONS.REQUEST: {
        processingRequestsNumber += 1;
        break;
      }
      case DELETE_ACTIONS.GOT:
      case DELETE_ACTIONS.ERROR:
      case ADD_ACTIONS.GOT:
      case ADD_ACTIONS.ERROR: {
        processingRequestsNumber -= 1;
        if (processingRequestsNumber === 0) {
          // all requests are completed

          // so we send request to get fresh todo list
          yield put(req());

          // clear deleting ids only after fresh todo list was requested
          // it is needed to disable items in right way and
          // avoid repeated deleting requests
          yield put(clearDeletingIds());
        }
      }
    }
  }
}
```

**Comment by Andarist** [comment]
I didnt mean tracking those completions with state/reducers, but in the saga itself (as some local variable). You have done exactly that 😉 If this code solves ur case I dont see anything wrong with it.

**Comment by hariDasu** [comment]
what if i'm asynchronously waiting for the success of  three different calls?

    const floorTilesSuccess = yield take("UNZIP_FLOOR_TILES_SUCCESS");
    const imagesSuccess = yield take("UNZIP_IMAGES_SUCCESS");
    const docsSuccess = yield take("UNZIP_DOCS_SUCCESS");
